### PR TITLE
AO3-4159 - Add last login information to user 

### PR DESCRIPTION
**Pull request by zz9pzza**
https://code.google.com/p/otwarchive/issues/detail?id=4167
https://otwarchive.atlassian.net/browse/AO3-4159

### CODE REVIEW

**Comment by zz9pzza** [comment]
The reason I am storing them temporarily in redis is that if we changed the session key then everyone would need to log in at once and bring the database down in a collapse of writes, this has the set of writes wrapped in a transaction so they only count as one chunk of work.

It also means we could have points in the app where we have an active user check and not kill the database using the same method.

File: app/views/admin/admin_users/_user_history.html.erb [hunk 1]
```diff
@@ -13,17 +13,27 @@
       <% unless @log_items.blank?
         @log_items.each do |item| %>
         <tr>
-          <td><%= item.created_at %></td>
+          <td><%= item.created_at %> ( <abbr class="timeago" title="<%= item.created_at.iso8601 =%>"><%= item.created_at.iso8601  =%></abbr> )</td>
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: Can you give me some example output for this? That is, will it look like one of these? 
`<td>2014-11-03 22:53:47 -0500 (<abbr class="timeago" title="2 months ago">2 months ago</abbr>)</td>`

*`<td>2014-11-03 22:53:47 -0500 (<abbr class="timeago" title="2 months ago">2 months ago</abbr>)</td>`

Also, I think you have might have too many `=` signs -- it's usually `<%= item.created_at.iso8601 %>` -- but if the second `=` is required for the iso method to work, just ignore me.

File: app/views/admin/admin_users/_user_history.html.erb [hunk 2]
```diff
@@ -13,17 +13,27 @@
...
         <% end %>
       <% end %>
       <tr>
-        <td><%= @user.created_at %></td>
+        <td><%= @user.created_at %> ( <abbr class="timeago" title="<%= @user.created_at.iso8601 =%>"><%= @user.created_at.iso8601  =%></abbr> )</td>
         <td><%= ts('Account Created') %></td>
         <td><%= ts('System Generated') %></td>
       </tr>
+      <tr>
+        <td><% if @user.last_sign_in_at.nil? %><%= ts('No logins recorded') %><%else%><%= @user.last_sign_in_at %> ( <abbr class="timeago" title="<%= @user.last_sign_in_at.iso8601 =%>"><%= @user.last_sign_in_at.iso8601 =%></abbr> ) <%end%> </td>
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: Could you break this up onto multiple lines so it's easier to read? More like:

```
<td>
  <% if @user.last_sign_in_at.nil? %>
    <%= ts('No logins recorded') %>
  <% else %>
    <%= @user.last_sign_in_at %> (<abbr class="timeago" title="<%= @user.last_sign_in_at.iso8601 =%>"><%= @user.last_sign_in_at.iso8601 =%></abbr>)
  <% end %>
</td>
```

File: app/views/admin/admin_users/_user_history.html.erb [hunk 3]
```diff
@@ -13,17 +13,27 @@
...
+        <td><%= ts('Last Login') %></td>
+        <td><%= ts('System Generated') %></td>
+      </tr>
+      <tr>
+        <td><% if @user.last_active_at.nil? %><%= ts('No activity recorded') %><%else%><%= @user.last_active_at %> ( <abbr class="timeago" title="<%= @user.last_active_at.iso8601 =%>"><%= @user.last_active_at.iso8601 =%></abbr> ) <%end%> </td>
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: Could you break this up onto multiple lines as well? Like:

```
<td>
  <% if @user.last_active_at.nil? %>
    <%= ts('No activity recorded') %>
  <% else %>
    <%= @user.last_active_at %> (<abbr class="timeago" title="<%= @user.last_active_at.iso8601 =%>"><%= @user.last_active_at.iso8601 =%></abbr>)
  <% end %>
</td>
```

File: app/views/admin/admin_users/show.html.erb [hunk 4]
```diff
@@ -56,7 +56,22 @@
 
   <p class="submit"><%= submit_tag ts("Update") %></p>
 <% end %>
-
 <%= render :partial => 'user_history', :locals => {:user => @user} %>
 <!--/content-->
 </div>
+<script src="//ajax.googleapis.com/ajax/libs/jquery/1.9.0/jquery.min.js" type="text/javascript"></script>
```
Comment (sarken) [inline app/views/admin/admin_users/show.html.erb]: We already load jQuery and jQuery UI on all our pages, so you'll need to remove lines 62-71. In order to ensure that the jquery.timeago.min.js script loads after jQuery, you _could_ do this:

```
<%= content_for :footer_js do %>
  <script src="/javascripts/jquery.timeago.min.js" type="text/javascript"></script>
  <script type="text/javascript">
    $(document).ready(function() {
      $("abbr.timeago").timeago();
   });
  </script>
<% end>
```
However, it's a pretty useful script that I can easily see us wanting to use elsewhere on the site, so it would be even better to add it directly to the /views/layouts/_javascripts.html.erb file: `<%= javascript_include_tag "jquery.scrollTo.min", "jquery.livequery.min", "rails", "application", "bootstrap/bootstrap-dropdown.min", "jquery-shuffle", "jquery.tokeninput.min", "jquery.trap.min", "ao3modal.min", "jquery.cookie", "jquery.timeago.min" %>`

Then you can put `$j('.timeago').timeago();` in the topmost section of /public/javascripts/application.js: 

```
$j(document).ready(function() {
    setupToggled();
    if ($j('#work-form')) { hideFormFields(); };
    hideHideMe();
    showShowMe();
    handlePopUps();
    attachCharacterCounters();
    setupAccordion();
    setupDropdown();

    // remove final comma from comma lists in older browsers
    $j('.commas li:last-child').addClass('last');

    // make Share buttons on works and own bookmarks visible
    $j('.actions').children('.share').removeClass('hidden');

    // convert an absolute time to a relative time, e.g. 1 day ago
    $j('.timeago').timeago();

    prepareDeleteLinks();
    thermometer();
});
```

(You don't need to use `abbr.timeago` because we'll never use the class name `timeago` without intending to call this script.)

**Comment by zz9pzza** [comment]
Thanks for those sarken, I have actioned them.

File: app/views/admin/admin_users/_user_history.html.erb [hunk 5]
```diff
@@ -13,17 +13,39 @@
...
+        </td>
+        <td><%= ts('Last Login') %></td>
+        <td><%= ts('System Generated') %></td>
+      </tr>
+      <tr>
+        <td>
+            <% if @user.last_active_at.nil? %>
+               <%= ts('No activity recorded') %>
+            <%else%>
+               <%= @user.last_active_at %> ( <abbr class="timeago" title="<%= @user.last_active_at =%>"><%= @user.last_active_at.iso8601 =%></abbr> )
```
Comment (CristinaRO) [inline app/views/admin/admin_users/_user_history.html.erb]: That last equal sign shouldn't be there... There's no `=%>` Also, would you update the pull request so it goes through the Hound's teeth, please? :D

File: app/views/admin/admin_users/_user_history.html.erb [hunk 6]
```diff
@@ -13,17 +13,39 @@
       <% unless @log_items.blank?
         @log_items.each do |item| %>
         <tr>
-          <td><%= item.created_at %></td>
+          <td><%= item.created_at %> ( <abbr class="timeago" title="<%= item.created_at %>"><%= item.created_at.iso8601  %></abbr> )</td>
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: There shouldn't be a space between the parentheses as the `<abbr>` tags. It should be `(<abbr class="timeago" title="<%= item.created_at %>"><%= item.created_at.iso8601 %></abbr>)`. There's also an extra space in `<%= item.created_at.iso8601  %>` that should be removed.

File: app/views/admin/admin_users/_user_history.html.erb [hunk 7]
```diff
@@ -13,17 +13,39 @@
...
           <td><%= log_item_action_name(item.action) %><%= item.role.name if item.role %><%= item.enddate if item.enddate %></td>
           <td><%= item.note %></td>
         </tr>
         <% end %>
       <% end %>
       <tr>
-        <td><%= @user.created_at %></td>
+        <td><%= @user.created_at %> ( <abbr class="timeago" title="<%= @user.created_at %>"><%= @user.created_at.iso8601  %></abbr> )</td>
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: There shouldn't be a space between the parentheses as the `<abbr>` tags. It should be `(<abbr class="timeago" title="<%= @user.created_at %>"><%= @user.created_at.iso8601 %></abbr>)`. There's also an extra space in `<%= @user.created_at.iso8601  %>` that should be removed.

File: app/views/admin/admin_users/_user_history.html.erb [hunk 8]
```diff
@@ -13,17 +13,39 @@
...
         <td><%= ts('Account Created') %></td>
         <td><%= ts('System Generated') %></td>
       </tr>
+      <tr>
+        <td>
+            <% if @user.last_sign_in_at.nil? %>
+               <%= ts('No logins recorded') %>
+            <%else%>
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: Should be `<% else %>` with a single space on either side of the word.

File: app/views/admin/admin_users/_user_history.html.erb [hunk 9]
```diff
@@ -13,17 +13,39 @@
...
+               <%= @user.last_sign_in_at %> ( <abbr class="timeago" title="<%= @user.last_sign_in_at %>"><%= @user.last_sign_in_at.iso8601 %></abbr> ) 
+            <%end%>
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: Should be `<% end %>`

File: app/views/admin/admin_users/_user_history.html.erb [hunk 10]
```diff
@@ -13,17 +13,39 @@
...
+        <td><%= @user.created_at %> ( <abbr class="timeago" title="<%= @user.created_at %>"><%= @user.created_at.iso8601  %></abbr> )</td>
         <td><%= ts('Account Created') %></td>
         <td><%= ts('System Generated') %></td>
       </tr>
+      <tr>
+        <td>
+            <% if @user.last_sign_in_at.nil? %>
+               <%= ts('No logins recorded') %>
+            <%else%>
+               <%= @user.last_sign_in_at %> ( <abbr class="timeago" title="<%= @user.last_sign_in_at %>"><%= @user.last_sign_in_at.iso8601 %></abbr> )
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: There shouldn't be spaces between the parentheses and the `<abbr>`. It should be `(<abbr class="timeago" title="<%= @user.last_sign_in_at %>"><%= @user.last_sign_in_at.iso8601 %></abbr>)`

File: app/views/admin/admin_users/_user_history.html.erb [hunk 11]
```diff
@@ -13,17 +13,39 @@
...
+            <%else%>
+               <%= @user.last_sign_in_at %> ( <abbr class="timeago" title="<%= @user.last_sign_in_at %>"><%= @user.last_sign_in_at.iso8601 %></abbr> ) 
+            <%end%>
+        </td>
+        <td><%= ts('Last Login') %></td>
+        <td><%= ts('System Generated') %></td>
+      </tr>
+      <tr>
+        <td>
+            <% if @user.last_active_at.nil? %>
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: The indenting here in this `<td>` is a little overzealous -- it should be two spaces, not four.

File: app/views/admin/admin_users/_user_history.html.erb [hunk 12]
```diff
@@ -13,17 +13,39 @@
...
+               <%= ts('No activity recorded') %>
+            <%else%>
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: Should be `<% else %>`

File: app/views/admin/admin_users/_user_history.html.erb [hunk 13]
```diff
@@ -13,17 +13,39 @@
...
+               <%= @user.last_active_at %> ( <abbr class="timeago" title="<%= @user.last_active_at =%>"><%= @user.last_active_at.iso8601 %></abbr> ) 
+            <%end%>
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: Should be `<% end %>`

File: app/views/admin/admin_users/_user_history.html.erb [hunk 14]
```diff
@@ -13,17 +13,39 @@
...
+        </td>
+        <td><%= ts('Last Login') %></td>
+        <td><%= ts('System Generated') %></td>
+      </tr>
+      <tr>
+        <td>
+            <% if @user.last_active_at.nil? %>
+               <%= ts('No activity recorded') %>
+            <%else%>
+               <%= @user.last_active_at %> ( <abbr class="timeago" title="<%= @user.last_active_at =%>"><%= @user.last_active_at.iso8601 %></abbr> )
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: Stop me if you've heard this one before, but there shouldn't be spaces between the parentheses and the `<abbr>` tag. :P It should be `(<abbr class="timeago" title="<%= @user.last_active_at %>"><%= @user.last_active_at.iso8601 %></abbr>)`. There's also a stray `=` in `<%= @user.last_active_at =%>`.

**Comment by sarken** [comment]
Unfortunately, this is no longer automatically mergeable. No, rush, though!

**Comment by scottsds** [comment]
Sorry @zz9pzza this went grey again.

File: app/views/admin/admin_users/_user_history.html.erb (hunk 11 above)
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: Just reiterating the above comment.

File: app/views/admin/admin_users/_user_history.html.erb [hunk 15]
```diff
@@ -13,17 +13,39 @@
...
+        </td>
+        <td><%= ts('Last Login') %></td>
+        <td><%= ts('System Generated') %></td>
+      </tr>
+      <tr>
+        <td>
+          <% if @user.last_active_at.nil? %>
+            <%= ts('No activity recorded') %>
+          <% else %>
+            <%= @user.last_active_at %> (<abbr class="timeago" title="<%= @user.last_active_at =%>"><%= @user.last_active_at.iso8601 %></abbr>)
```
Comment (sarken) [inline app/views/admin/admin_users/_user_history.html.erb]: `<%= @user.last_active_at =%>` -- I don't think we want that second `=` at the end

File: db/migrate/20150101133905_add_last_login_time_to_users.rb [hunk 16]
```diff
@@ -0,0 +1,11 @@
+class AddLastLoginTimeToUsers < ActiveRecord::Migration
+  def self.up
+    add_column :users, :last_sign_in_at, :datetime
```
Comment (sarken) [inline db/migrate/20150101133905_add_last_login_time_to_users.rb]: I was wondering, why is the database column called `last_sign_in_at` rather than `last_login_at` (or `last_log_in_at`), since most of the code seems to use `'login` rather than `sign_in`? If there's a reason it has to be that way, that's cool! But if there's not, it would probably be less confusing for future coders if it consistently used one term.

**Comment by shalott** [comment]
This looks good to me but has no automated tests at all? Would suggest this feature (can help with step definitions if you need):

```
Scenario: User's last login date is saved
Given the user "new_user" exists and is activated
  And I am logged in as an admin
When I go to the abuse administration page for "new_user"
Then I should see "No logins recorded"
When I am logged in as "new_user"
  And the last login information has been updated
  And I am logged in as an admin
  And I go to the abuse administration page for "new_user"
Then I should not see "No logins recorded"
  And I should see the last login time for "new_user"
```

**Comment by sarken** [comment]
No rush, but when you get the chance, can you make a newer version of this migration so we don't run into any issues? Thanks!

**Comment by cesy** [comment]
The recommended tests are still missing, and a few other comments from hound.

**Comment by cesy** [comment]
2 new valid comments from Hound to fix. I've looked through the actual code but it needs someone more familiar with Redis than me to review.

File: app/models/user.rb [hunk 17]
```diff
@@ -532,4 +538,33 @@ def update_pseud_name
...
+
+  def perform(operation)
+    case operation
+    when 'update_last_login'
+      self.update_last_login
+    end
+  end
+
+  # called from rake
+  def self.update_last_login
```
Comment (shalott) [inline app/models/user.rb]: You don't want this to be private since it's called from rake so would suggest moving this method somewhere else in the file out from under the private declaration.

Ideally tbh this should go into a separate module and not be in the already horribly bloated user.rb class.

File: features/step_definitions/admin_steps.rb [hunk 18]
```diff
@@ -340,3 +340,11 @@ def update_settings(settings)
   u = User.find_by_login(user)
   u.update_attribute(:banned, false)
 end
+
+When(/^the last login information has been updated$/) do
+  User.update_last_login
+end
+
+Then(/^I should see the last login time/) do
+  step('I should not see "No logins recorded"')
```
Comment (shalott) [inline features/step_definitions/admin_steps.rb]: Here I think you want something a bit more specific and also this step doesn't actually match completely.  

```
Then /^I should see the last login time for "([^"]*)"$/ do |user|
  u = User.find_by_login(user)
  assert u.last_active_at.match(//) # not sure what the format is but eg the current year or something
  step %{I should see "#{u.last_active_at}"}
end
```

which will actually make sure that the value is what you expect and that it is displayed.

File: app/controllers/user_sessions_controller.rb [hunk 19]
```diff
@@ -22,6 +22,8 @@ def create
       if @user_session.save
         flash[:notice] = ts("Successfully logged in.")
         @current_user = @user_session.record
+        REDIS_GENERAL.sadd("last_login_list", params[:user_session][:login])
+        REDIS_GENERAL.set("last_login_#{params[:user_session][:login]}", Time.now.utc)
```
Comment (shalott) [inline app/controllers/user_sessions_controller.rb]: Sorry I have had another thought reading this first thing in the morning with freshly caffeinated brain: shouldn't we use user id instead? Then if a user changes their name we still have their correct last login time.

File: app/models/user.rb [hunk 20]
```diff
@@ -540,4 +546,33 @@ def update_pseud_name
...
+    User.transaction do
+      list = REDIS_GENERAL.smembers("last_login_list").each
+      list.each do |username|
```
Comment (shalott) [inline app/models/user.rb]: would change this to userid throughout for clarity and make it user = User.find(userid) instead but otherwise I don't think the code needs to be changed in any way

**Comment by sarken** [comment]
I just want to mention Devise (#2505) adds the following:

```
     add_column :users, :sign_in_count, :integer, default: 0, null: false
     add_column :users, :current_sign_in_at, :datetime
     add_column :users, :last_sign_in_at, :datetime
     add_column :users, :current_sign_in_ip, :string
     add_column :users, :last_sign_in_ip, :string
```

So we might only need to change the admin_users/_user_history.html.erb view to show this information once Devise is merged.

**Comment by redsummernight** [comment]
Devise module Trackable adds `last_sign_in_at`, so after #3378 we only need to expose that to the UI.

**Comment by zz9pzza** [comment]
Replaced by https://github.com/otwcode/otwarchive/pull/3545
